<a href="https://colab.research.google.com/github/Aayushi-Sharma24/EDA-Hotel-Booking-Analysis/blob/main/Hotel_Booking_Analysis_Capstone_Project_Final.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## <b> Have you ever wondered when the best time of year to book a hotel room is? Or the optimal length of stay in order to get the best daily rate? What if you wanted to predict whether or not a hotel was likely to receive a disproportionately high number of special requests? This hotel booking dataset can help you explore those questions!

## <b>This data set contains booking information for a city hotel and a resort hotel, and includes information such as when the booking was made, length of stay, the number of adults, children, and/or babies, and the number of available parking spaces, among other things. All personally identifying information has been removed from the data. </b>

## <b> Explore and analyze the data to discover important factors that govern the bookings. </b>

#Introduction
###Hotel industry is very volatile industry and the bookings depend on variety of factors such as type of hotels, seasonality, days of week and many more. This make analyzing the patterns available in the past data more important to help hotels plan better.

#Let's have a look at Column Information
**Hotel**: types of hotel

**is_canceled**: booking canceled or not

**lead_time**: No. of days before actual arrival in the hotel

**arrival_date_year**: year of booking

**arrival_date_month**: month of booking

**arrival_date_week_number**: week number of the year of booking

**arrival_date_day_of_month**: arrival month date

**stays_in_weekend_nights**: no. of weekends guests stayed

**stays_in_week_nights**: no. of weekdays guests stayed

**adults**: No. of Adults

**children**: No. of children

**babies**: No. of babaies

**meal**: Type of meal booked. 

*   Undefined/SC – no meal package; 
*   BB – Bed & Breakfast;
*   HB – Half board (breakfast and one other meal);
*   FB – Full board (breakfast, lunch and dinner)

**country**: country from which guests belong

**market_segment**: group of people who share one or more common characteristics, lumped together for marketing purposes

*   TA: Travel agents
*   TO: Tour operators

**distribution_channel**: chain of businesses

*   TA: Travel agents
*   TO: Tour operators

**is_repeated_guest**:  booking name was from repeated guest or not

**previous_cancellations**:  No. of previous bookings that were cancelled by the customer prior to the current booking

**previous_bookings_not_canceled**:  No. of previous bookings not cancelled by the customer prior to the current booking

**reserved_room_type**:  Code of room type reserved

**assigned_room_type**:  Code for the type of room assigned to the booking

**booking_changes**:  No. of changes made to the booking

**deposit_type**: 
*   Deposit – no deposit was made;
*   Non Refund 
*   Refundable

**agent**: ID of the travel agency that made the booking

**company**: ID of the company that made the booking

**day_in_waiting_list**: No. of days the booking was in the waiting list before it was confirmed to the customer

**customer_type**: type of customer

**adr (average daily rate)**

**required_car_parking_spaces**: No. of car parking spaces required by the customer

**total_of_special_requests**: No. of special requests made by the customer 

**reservation_status**: 
*   Canceled – booking was canceled by the customer;
*   Check-Out – customer has checked in but already departed;
*   No-Show – customer did not check-in 

**reservation_status_date**: Date at which the last status was set

#Objective
### To create an indepth analysis to figure out the standard patterns of booking based on various factors

# Data Preparation and Cleaning

### Let's import the required libraries that we will be using for this project

In [1]:
# Importing required packages
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns

## Importing dataset into notebook

In [2]:
# Mount drive
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


###Let's Download the dataset to use it as a pandas dataframe

In [16]:
#Loading csv file to dataframe
path = "/content/drive/MyDrive/Colab Notebooks/project EDA/Hotel Bookings.csv"
df = pd.read_csv(path)

In [17]:
df.head()

hotel  is_canceled  ...  reservation_status  reservation_status_date
0  Resort Hotel            0  ...           Check-Out               2015-07-01
1  Resort Hotel            0  ...           Check-Out               2015-07-01
2  Resort Hotel            0  ...           Check-Out               2015-07-02
3  Resort Hotel            0  ...           Check-Out               2015-07-02
4  Resort Hotel            0  ...           Check-Out               2015-07-03

[5 rows x 32 columns]

##Let's Explore the DataFrame
###Looking into shape of data to find out number of rows and columns

In [6]:
# checks the no. of rows and columns
df.shape

(119390, 32)

 So there are 119390 rows and 32 columns in our DataFrame

### Let's have  look at all the column names present in our DataFrame

In [7]:
# Check column names
df.columns

Index(['hotel', 'is_canceled', 'lead_time', 'arrival_date_year',
       'arrival_date_month', 'arrival_date_week_number',
       'arrival_date_day_of_month', 'stays_in_weekend_nights',
       'stays_in_week_nights', 'adults', 'children', 'babies', 'meal',
       'country', 'market_segment', 'distribution_channel',
       'is_repeated_guest', 'previous_cancellations',
       'previous_bookings_not_canceled', 'reserved_room_type',
       'assigned_room_type', 'booking_changes', 'deposit_type', 'agent',
       'company', 'days_in_waiting_list', 'customer_type', 'adr',
       'required_car_parking_spaces', 'total_of_special_requests',
       'reservation_status', 'reservation_status_date'],
      dtype='object')

 ### Looking into the datatype of each column

In [9]:
# check the datatypes of each column
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119390 entries, 0 to 119389
Data columns (total 32 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   hotel                           119390 non-null  object 
 1   is_canceled                     119390 non-null  int64  
 2   lead_time                       119390 non-null  int64  
 3   arrival_date_year               119390 non-null  int64  
 4   arrival_date_month              119390 non-null  object 
 5   arrival_date_week_number        119390 non-null  int64  
 6   arrival_date_day_of_month       119390 non-null  int64  
 7   stays_in_weekend_nights         119390 non-null  int64  
 8   stays_in_week_nights            119390 non-null  int64  
 9   adults                          119390 non-null  int64  
 10  children                        119386 non-null  float64
 11  babies                          119390 non-null  int64  
 12  meal            

# Dealing with Missing values

### Let's check how many cells have Null values in our DataFrame

In [10]:
# Checking Null values
df.isnull().sum()

hotel                                  0
is_canceled                            0
lead_time                              0
arrival_date_year                      0
arrival_date_month                     0
arrival_date_week_number               0
arrival_date_day_of_month              0
stays_in_weekend_nights                0
stays_in_week_nights                   0
adults                                 0
children                               4
babies                                 0
meal                                   0
country                              488
market_segment                         0
distribution_channel                   0
is_repeated_guest                      0
previous_cancellations                 0
previous_bookings_not_canceled         0
reserved_room_type                     0
assigned_room_type                     0
booking_changes                        0
deposit_type                           0
agent                              16340
company         

We can see that we have 4 columns with missing values.
### Lets check these values as percentages.

In [11]:
# Checking Null values percentage
df.isnull().sum()/len(df.index)*100

hotel                              0.000000
is_canceled                        0.000000
lead_time                          0.000000
arrival_date_year                  0.000000
arrival_date_month                 0.000000
arrival_date_week_number           0.000000
arrival_date_day_of_month          0.000000
stays_in_weekend_nights            0.000000
stays_in_week_nights               0.000000
adults                             0.000000
children                           0.003350
babies                             0.000000
meal                               0.000000
country                            0.408744
market_segment                     0.000000
distribution_channel               0.000000
is_repeated_guest                  0.000000
previous_cancellations             0.000000
previous_bookings_not_canceled     0.000000
reserved_room_type                 0.000000
assigned_room_type                 0.000000
booking_changes                    0.000000
deposit_type                    

We can see that approx **94%** of values in "**company**" column and **13%** of values in "**agent**" column are Null. Such huge number of values are Null in these column which can create issue while analyzing data, hence we will delete these columns

The columns **“children”** and **“country”** have a low percentage of missing values. So, we will remove the full row on missing cells.

In [18]:
# Deleting agent and company column
df = df.drop(columns = ['agent', 'company'])

# check number of rows and column
df.shape

(119390, 30)

In [19]:
# check all column names and shape
df.columns

Index(['hotel', 'is_canceled', 'lead_time', 'arrival_date_year',
       'arrival_date_month', 'arrival_date_week_number',
       'arrival_date_day_of_month', 'stays_in_weekend_nights',
       'stays_in_week_nights', 'adults', 'children', 'babies', 'meal',
       'country', 'market_segment', 'distribution_channel',
       'is_repeated_guest', 'previous_cancellations',
       'previous_bookings_not_canceled', 'reserved_room_type',
       'assigned_room_type', 'booking_changes', 'deposit_type',
       'days_in_waiting_list', 'customer_type', 'adr',
       'required_car_parking_spaces', 'total_of_special_requests',
       'reservation_status', 'reservation_status_date'],
      dtype='object')

Column "company" and "agent" are deleted

Now, deleting rows with Null values in "children" and "country" column

In [20]:
 # Deleting rows with empty cells
df = df.dropna(axis = 0)

### Let's check again for missing values

In [21]:
# Checking Null values
df.isnull().sum()

hotel                             0
is_canceled                       0
lead_time                         0
arrival_date_year                 0
arrival_date_month                0
arrival_date_week_number          0
arrival_date_day_of_month         0
stays_in_weekend_nights           0
stays_in_week_nights              0
adults                            0
children                          0
babies                            0
meal                              0
country                           0
market_segment                    0
distribution_channel              0
is_repeated_guest                 0
previous_cancellations            0
previous_bookings_not_canceled    0
reserved_room_type                0
assigned_room_type                0
booking_changes                   0
deposit_type                      0
days_in_waiting_list              0
customer_type                     0
adr                               0
required_car_parking_spaces       0
total_of_special_requests   

Now, there are no missing values in our DataFrame